Add new SelectelProvider class to support DNS v2 API.

dchudik opened the pull request:
Soon we are going to publicly announce that v2 DNS API is ready to use. API is available from internet, but there is no public docs, UI etc.
DNS v1 will be available and maintained but we gradually limit access to it until full shutdown.
So we would like to support both versions with this plugin.

This commit is all about that and also minor changes.

Changes
* Move existing provider and related tests to separate directories `v1`
* Rename public name from `SelectelProvider` to `SelectelProviderLegacy`
* Add `list_zones()` method to support "*"  for planning
* Fix SSHFP parsing bug, caused by trailing dot in fingerprint which lead to constant re-update of record
* Move version varible to separate file, since now it is utilized by two providers. Storing it in `__init__.py` causes cycling imports
* Update `script/release` and `setup.py` to parse version from another location
* Update `readme.md` with focus on new provider

New
* Add new `SelectelProvider` class to support v2 API
* Isolate API calls into separate class
* Add tests for SelectelProvider

After these changes we also planning to bump plugin version to 1.0.0
All examples in readme are working(just change to unique zone-name). So if there is need to check it, they should suffice.

dchudik committed:
Add new SelectelProvider class to support DNS v2 API.

ross commented:
Looks like `match` is failing things in 3.8/3.9:

https://github.com/octodns/octodns-selectel/actions/runs/7615066154/job/20792848891?pr=40#step:4:270

dchudik committed:
Replaced match-case to if-else for supporting python3.8. Rename fields in rrset and zone.

dchudik commented:
Replaced match-case to if-else in `octodns_selectel/v2/dns_client.py` for supporting python3.8/python3.9.

Renamed `uuid` to `id` in zone and rrset .
Renamed `zone` to `zone_id` in rrset.
API renamed these fields same.

ross commented:
Not really sure what it doesn't like now, I haven't done much w/types.

```
    def _request_all_entities(self, path, offset=0) -> list[int]:
E   TypeError: 'type' object is not subscriptable
```

https://github.com/octodns/octodns-selectel/actions/runs/7646141503/job/20870444134?pr=40#step:4:285

dchudik committed:
Removed type annotation for working with python3.8. Removed match-case to if-else in mappings

dchudik commented:
Removed type annotation for function.
Replace match-case to if-else in `octodns_selectel/v2/mappings.py`.

ross reviewed (approved):
So this is obviously a huge change and given that I don't have access to or deep knowledge of selecttel there's only so much that I can do in terms of the review details. I've skimmed over things and nothing that I saw jumped out. If you have a coworker or someone that can do a more focused look over things that's probably a good idea. 

I'll give it until Tuesday next week or there about to allow for anyone to weight in before merging.

Archirk commented:
> So this is obviously a huge change and given that I don't have access to or deep knowledge of selecttel there's only so much that I can do in terms of the review details. I've skimmed over things and nothing that I saw jumped out. If you have a coworker or someone that can do a more focused look over things that's probably a good idea.
> 
> I'll give it until Tuesday next week or there about to allow for anyone to weight in before merging.

Hello,
This pull request has been done under my supervision, as his colleague at Selectel. It has been reviewed multiple times.
Core functionality has been tested and works fine.
After public launch we may eventually run in and do some bug fixes. But as for upgrading to major version - it safe to do.

ross commented:
> This pull request has been done under my supervision, as his colleague at Selectel. It has been reviewed multiple times.
> Core functionality has been tested and works fine.

Great. Thanks. 


Just realized that a CHANGELOG.md entry wasn't included here. Definitely something that belongs there since the default behavior is going to be switching to the v2 api. Will approve & merge as soon as I see that come through. 

I assume you all would like a release cut after that?

Archirk commented:
> Just realized that a CHANGELOG.md entry wasn't included here. Definitely something that belongs there since the default behavior is going to be switching to the v2 api. Will approve & merge as soon as I see that come through.
> 
> I assume you all would like a release cut after that?

If i remeber correctly, you usually merge/accept pull request first, than do commit with chagelog and version update and tag. So we didn't update changelog and version inside module  intentionally.

I hope that `changes` and `new` paragraphs from PR are suitable to be just copy&paste into changelog.

ross commented:
> If i remeber correctly, you usually merge/accept pull request first, than do commit with chagelog and version update and tag. So we didn't update changelog and version inside module intentionally.

Yeah. For releases the changelog date is finalized and the version number is bumped. 

I generally try to have the changelog entries happen with the PRs that make the change so that blame on the CHANGELOG will take you to the PR that made the relevant change. It's for sure not 100% of the time, but it's a goal.

> I hope that `changes` and `new` paragraphs from PR are suitable to be just copy&paste into changelog.

Looks like they probably are.

dchudik committed:
Update changelog. Fixes for idna zones.

dchudik commented:
Updated changelog and added some fixes for idna zones. 
I there anything else that is needed to be done you now can merge, bump and tag commit?

ross commented on CHANGELOG.md:
```diff
@@ -1,3 +1,21 @@
+## v1.0.0 - 2024-02-01 - Support for DNSv2 API and structural project changes
```
Don't want to call something w/changes this large 1.0/stable. Can do 0.99/beta to allow for patches and if things are stable for a bit it can be bumped to 1.0

```suggestion
## v0.99.0 - 2024-02-01 - Support for DNSv2 API and structural project changes
```

ross committed:
beta version rather than 1.0

ross commented:
/cc https://github.com/octodns/octodns-selectel/pull/41 for release
